File: frontend/types/inventory.ts
import { z } from "zod";
import { GasTypeSchema } from "./common";

export const InventoryAdjustmentSchema = z
  .object({
    id: z.string(),
    group_id: z.string().nullish(),
    gas_type: GasTypeSchema,
    delta_full: z.number(),
    delta_empty: z.number(),
    reason: z.string().nullish(),
    effective_at: z.string(),
    created_at: z.string(),
    is_deleted: z.boolean().optional(),
  })
  .passthrough();
export type InventoryAdjustment = z.infer<typeof InventoryAdjustmentSchema>;

export const InventoryRefillSummarySchema = z
  .object({
    refill_id: z.string(),
    date: z.string(),
    time_of_day: z.enum(["morning", "evening"]).optional(),
    effective_at: z.string(),
    buy12: z.number(),
    return12: z.number(),
    buy48: z.number(),
    return48: z.number(),
    new12: z.number().optional(),
    new48: z.number().optional(),
    debt_cash: z.number().optional().default(0),
    debt_cylinders_12: z.number().optional().default(0),
    debt_cylinders_48: z.number().optional().default(0),
    is_deleted: z.boolean().optional(),
    deleted_at: z.string().nullish(),
  })
  .passthrough();
export type InventoryRefillSummary = z.infer<typeof InventoryRefillSummarySchema>;

export const InventoryAdjustmentUpdateSchema = z.object({
  delta_full: z.number().optional(),
  delta_empty: z.number().optional(),
  reason: z.string().optional(),
  note: z.string().optional(),
  allow_negative: z.boolean().optional(),
});
export type InventoryAdjustmentUpdate = z.infer<typeof InventoryAdjustmentUpdateSchema>;

export const InventorySnapshotSchema = z.object({
  as_of: z.string(),
  full12: z.number(),
  empty12: z.number(),
  total12: z.number(),
  full48: z.number(),
  empty48: z.number(),
  total48: z.number(),
  reason: z.string().nullish(),
});
export type InventorySnapshot = z.infer<typeof InventorySnapshotSchema>;

export const InventoryRefillDetailsSchema = z.object({
  refill_id: z.string(),
  business_date: z.string(),
  time_of_day: z.enum(["morning", "evening"]).optional(),
  effective_at: z.string(),
  buy12: z.number(),
  return12: z.number(),
  buy48: z.number(),
  return48: z.number(),
  total_cost: z.number(),
  paid_now: z.number(),
  new12: z.number().optional(),
  new48: z.number().optional(),
  debt_cash: z.number().optional().default(0),
  debt_cylinders_12: z.number().optional().default(0),
  debt_cylinders_48: z.number().optional().default(0),
  notes: z.string().nullish(),
  before_full_12: z.number().optional(),
  before_empty_12: z.number().optional(),
  after_full_12: z.number().optional(),
  after_empty_12: z.number().optional(),
  before_full_48: z.number().optional(),
  before_empty_48: z.number().optional(),
  after_full_48: z.number().optional(),
  after_empty_48: z.number().optional(),
  is_deleted: z.boolean().optional(),
  deleted_at: z.string().nullish(),
});
export type InventoryRefillDetails = z.infer<typeof InventoryRefillDetailsSchema>;

File: backend/app/schemas/inventory.py
from __future__ import annotations

from datetime import datetime
from typing import Literal, Optional

from pydantic import field_validator
from sqlmodel import SQLModel

from .common import GasType, InventoryAdjustReason, _non_negative


class InventoryAdjustCreate(SQLModel):
  happened_at: Optional[datetime] = None
  group_id: Optional[str] = None
  gas_type: GasType
  delta_full: int = 0
  delta_empty: int = 0
  reason: Optional[str] = None
  note: Optional[str] = None
  request_id: Optional[str] = None


class InventoryAdjustUpdate(SQLModel):
  delta_full: Optional[int] = None
  delta_empty: Optional[int] = None
  reason: Optional[str] = None
  note: Optional[str] = None


class InventoryAdjustmentRow(SQLModel):
  id: str
  group_id: Optional[str] = None
  gas_type: GasType
  delta_full: int
  delta_empty: int
  reason: Optional[str] = None
  effective_at: datetime
  created_at: datetime
  is_deleted: bool = False


class InventorySnapshot(SQLModel):
  as_of: datetime
  full12: int
  empty12: int
  total12: int
  full48: int
  empty48: int
  total48: int
  reason: Optional[str] = None


class InventoryInitCreate(SQLModel):
  date: Optional[str] = None
  full12: int = 0
  empty12: int = 0
  full48: int = 0
  empty48: int = 0
  reason: Optional[str] = None

  @field_validator("full12", "empty12", "full48", "empty48")
  @classmethod
  def _validate_non_negative(cls, value: Optional[int], info) -> Optional[int]:
    return _non_negative(value, info.field_name)


class InventoryRefillCreate(SQLModel):
  happened_at: Optional[datetime] = None
  buy12: int = 0
  return12: int = 0
  buy48: int = 0
  return48: int = 0
  total_cost: int = 0
  paid_now: int = 0
  debt_cash: int = 0
  debt_cylinders_12: int = 0
  debt_cylinders_48: int = 0
  note: Optional[str] = None
  new12: int = 0
  new48: int = 0
  request_id: Optional[str] = None

  @field_validator(
    "buy12",
    "return12",
    "buy48",
    "return48",
    "total_cost",
    "paid_now",
    "new12",
    "new48",
  )
  @classmethod
  def _validate_non_negative(cls, value: Optional[int], info) -> Optional[int]:
    return _non_negative(value, info.field_name)


class InventoryRefillSummary(SQLModel):
  refill_id: str
  date: str
  time_of_day: Optional[Literal["morning", "evening"]] = None
  effective_at: datetime
  created_at: Optional[datetime] = None
  buy12: int
  return12: int
  buy48: int
  return48: int
  new12: int = 0
  new48: int = 0
  debt_cash: int = 0
  debt_cylinders_12: int = 0
  debt_cylinders_48: int = 0
  live_debt_cash: Optional[int] = None
  live_debt_cylinders_12: Optional[int] = None
  live_debt_cylinders_48: Optional[int] = None
  is_deleted: bool = False
  deleted_at: Optional[datetime] = None
  kind: str = "refill"


class InventoryRefillUpdate(SQLModel):
  buy12: int = 0
  return12: int = 0
  buy48: int = 0
  return48: int = 0
  total_cost: int = 0
  paid_now: int = 0
  debt_cash: Optional[int] = None
  debt_cylinders_12: Optional[int] = None
  debt_cylinders_48: Optional[int] = None
  note: Optional[str] = None
  new12: int = 0
  new48: int = 0

  @field_validator(
    "buy12",
    "return12",
    "buy48",
    "return48",
    "total_cost",
    "paid_now",
    "new12",
    "new48",
  )
  @classmethod
  def _validate_non_negative(cls, value: Optional[int], info) -> Optional[int]:
    return _non_negative(value, info.field_name)


class InventoryRefillDetails(SQLModel):
  refill_id: str
  business_date: str
  time_of_day: Optional[Literal["morning", "evening"]] = None
  effective_at: datetime
  buy12: int
  return12: int
  buy48: int
  return48: int
  total_cost: int
  paid_now: int
  new12: int = 0
  new48: int = 0
  debt_cash: int = 0
  debt_cylinders_12: int = 0
  debt_cylinders_48: int = 0
  notes: Optional[str] = None
  is_deleted: bool = False
  deleted_at: Optional[datetime] = None
